File: ruby/test/test_basic.rb
# frozen_string_literal: true

require 'minitest/autorun'
require_relative '../lib/yosina'

class TestBasic < Minitest::Test
  def test_char_creation
    char = Yosina::Char.new(c: 'a', offset: 0)
    assert_equal 'a', char.c
    assert_equal 0, char.offset
    assert_nil char.source
  end

  def test_build_char_array
    result = Yosina::Chars.build_char_array('hello')
    assert_equal 6, result.length
    assert_equal 'h', result[0].c
    assert_equal 0, result[0].offset
    assert_equal 'o', result[4].c
    assert_equal 4, result[4].offset
  end

  def test_as_s
    chars = [
      Yosina::Char.new(c: 'h', offset: 0),
      Yosina::Char.new(c: 'e', offset: 1),
      Yosina::Char.new(c: 'l', offset: 2),
      Yosina::Char.new(c: 'l', offset: 3),
      Yosina::Char.new(c: 'o', offset: 4),
      Yosina::Char.new(c: '', offset: 5),
    ]
    result = Yosina::Chars.as_s(chars)
    assert_equal 'hello', result
  end

  def test_chars_extension
    chars = [
      Yosina::Char.new(c: 'h', offset: 0),
      Yosina::Char.new(c: 'e', offset: 1),
      Yosina::Char.new(c: 'l', offset: 2),
      Yosina::Char.new(c: 'l', offset: 3),
      Yosina::Char.new(c: 'o', offset: 4),
      Yosina::Char.new(c: '', offset: 5),
    ]
    class << chars
      include Yosina::Chars
    end
    assert_equal 'hello', chars.to_s
    assert_equal 6, chars.each.count
    assert_equal 'll', chars.select { |c| c.c == 'l' }.to_s
    assert_equal 'h-el-l-o-', chars.chunk { |c| c.offset % 3 != 0 }.map { |k, chunk| "#{chunk.to_s}-" }.join
    assert_equal 'olleh', chars.sort_by {|c| -c.offset }.to_s
    assert_equal({ 'h' => 'h', 'e' => 'e', 'l' => 'll', 'o' => 'o', '' => ''}, chars.group_by {|c| c.c }.transform_values(&:to_s))
  end

  def test_make_transliterator_with_space_normalization
    recipe = Yosina::TransliterationRecipe.new(replace_spaces: true)
    transliterator = Yosina.make_transliterator(recipe)

    # Test with ideographic space (should be normalized to regular space)
    test_input = "hello\u3000world" # ideographic space
    result = transliterator.call(test_input)
    assert_equal 'hello world', result
  end

  def test_make_transliterator_with_configs
    configs = [Yosina::TransliteratorConfig.new(:spaces)]
    transliterator = Yosina.make_transliterator(configs)

    # Test with ideographic space
    test_input = "hello\u3000world"
    result = transliterator.call(test_input)
    assert_equal 'hello world', result
  end

  def test_make_transliterator_with_string_configs
    configs = [:spaces]
    transliterator = Yosina.make_transliterator(configs)

    # Test with ideographic space
    test_input = "hello\u3000world"
    result = transliterator.call(test_input)
    assert_equal 'hello world', result
  end

  def test_invalid_transliterator_name
    assert_raises(Yosina::TransliteratorFactoryError) do
      Yosina.make_transliterator([:invalid_name])
    end
  end

  def test_chained_transliterators
    recipe = Yosina::TransliterationRecipe.new(
      replace_spaces: true,
      kanji_old_new: true
    )
    transliterator = Yosina.make_transliterator(recipe)

    # Test with ideographic space
    test_input = "hello\u3000world"
    result = transliterator.call(test_input)
    assert_equal 'hello world', result
  end
end

File: python/tests/test_basic.py
"""Basic functionality tests for yosina."""

import pytest

from yosina import make_transliterator
from yosina.chars import Char, build_char_list, from_chars
from yosina.recipes import TransliterationRecipe
from yosina.transliterators import TransliteratorConfig, TransliteratorIdentifier


def test_char_creation():
    """Test Char dataclass creation."""
    char = Char(c="a", offset=0)
    assert char.c == "a"
    assert char.offset == 0
    assert char.source is None


def test_build_char_list():
    """Test building character arrays from strings."""
    result = build_char_list("hello")
    assert len(result) == 6
    assert result[0].c == "h"
    assert result[0].offset == 0
    assert result[4].c == "o"
    assert result[4].offset == 4


def test_from_chars():
    """Test converting character arrays back to strings."""
    chars = [
        Char(c="h", offset=0),
        Char(c="e", offset=1),
        Char(c="l", offset=2),
        Char(c="l", offset=3),
        Char(c="o", offset=4),
    ]
    result = from_chars(chars)
    assert result == "hello"


def test_make_transliterator_with_empty_recipe():
    """Test creating a transliterator with an empty recipe."""
    recipe = TransliterationRecipe()

    # Empty recipe should raise an error since no transliterators are configured
    with pytest.raises(ValueError, match="at least one transliterator"):
        make_transliterator(recipe)


def test_make_transliterator_with_space_normalization():
    """Test space normalization functionality."""
    recipe = TransliterationRecipe(replace_spaces=True)
    transliterator = make_transliterator(recipe)

    # Test with ideographic space (should be normalized to regular space)
    test_input = "hello\u3000world"  # ideographic space
    result = transliterator(test_input)
    assert result == "hello world"


def test_recipe_mutual_exclusion():
    """Test that mutually exclusive options raise an error."""
    recipe = TransliterationRecipe(
        to_fullwidth=True,
        to_halfwidth=True,
    )

    with pytest.raises(ValueError, match="mutually exclusive"):
        make_transliterator(recipe)


def test_make_transliterator_with_configs():
    """Test creating a transliterator with direct configs."""
    configs: list[TransliteratorIdentifier | TransliteratorConfig] | TransliterationRecipe = [("spaces", {})]
    transliterator = make_transliterator(configs)

    # Test with ideographic space
    test_input = "hello\u3000world"
    result = transliterator(test_input)
    assert result == "hello world"


def test_empty_config_list_raises_error():
    """Test that empty config list raises an error."""
    with pytest.raises(ValueError, match="at least one transliterator"):
        make_transliterator([])


def test_invalid_transliterator_name():
    """Test that invalid transliterator names raise errors."""
    with pytest.raises(ValueError, match="transliterator not found"):
        make_transliterator([("invalid-name", {})])  # type: ignore
